File: internal/arc/merkle.go
package arc

import (
	"math/big"

	"github.com/consensys/gnark-crypto/ecc/bn254/fr"
	"github.com/consensys/gnark-crypto/ecc/bn254/fr/mimc"
)

// MiMCHash computes MiMC hash of two field elements.
// Matches the gnark circuit's mimcHash(left, right) exactly:
//
//	h.Write(left); h.Write(right); return h.Sum()
func MiMCHash(left, right *fr.Element) fr.Element {
	h := mimc.NewMiMC()
	lb := left.Bytes()
	rb := right.Bytes()
	h.Write(lb[:])
	h.Write(rb[:])
	sum := h.Sum(nil)
	var result fr.Element
	result.SetBytes(sum)
	return result
}

// MiMCHashInts computes MiMC hash of two int64 values.
func MiMCHashInts(a, b int64) fr.Element {
	var fa, fb fr.Element
	fa.SetInt64(a)
	fb.SetInt64(b)
	return MiMCHash(&fa, &fb)
}

// MerkleTree is a fixed-depth binary Merkle tree using MiMC-BN254.
type MerkleTree struct {
	Root   fr.Element
	Leaves []fr.Element
	Depth  int
	Layers [][]fr.Element
}

// ProofNode represents a sibling in a Merkle proof.
type ProofNode struct {
	Element fr.Element `json:"element"`
	// Index: 0 means the leaf is on the left, 1 means the leaf is on the right.
	// Matches gnark circuit PathIndices semantics.
	Index int `json:"index"`
}

// NewMerkleTree builds a fixed-depth binary Merkle tree from leaf hashes.
// Pads with zero hashes if fewer leaves than 2^depth.
func NewMerkleTree(leaves []fr.Element, depth int) *MerkleTree {
	maxLeaves := 1 << depth

	// Pad with zeros
	padded := make([]fr.Element, maxLeaves)
	copy(padded, leaves)
	// Remaining elements are zero-valued (fr.Element{} == 0)

	layers := make([][]fr.Element, depth+1)
	layers[0] = padded
	current := padded

	for level := 0; level < depth; level++ {
		next := make([]fr.Element, len(current)/2)
		for i := 0; i < len(current); i += 2 {
			next[i/2] = MiMCHash(&current[i], &current[i+1])
		}
		layers[level+1] = next
		current = next
	}

	return &MerkleTree{
		Root:   current[0],
		Leaves: padded,
		Depth:  depth,
		Layers: layers,
	}
}

// NewMerkleTreeFromEntries builds a tree from (key, value) pairs.
// Each leaf is mimcHash(key, value).
func NewMerkleTreeFromEntries(entries [][2]fr.Element, depth int) *MerkleTree {
	leaves := make([]fr.Element, len(entries))
	for i, entry := range entries {
		leaves[i] = MiMCHash(&entry[0], &entry[1])
	}
	return NewMerkleTree(leaves, depth)
}

// GetProof returns the Merkle proof for a leaf at the given index.
func (t *MerkleTree) GetProof(index int) []ProofNode {
	if index < 0 || index >= len(t.Leaves) {
		return nil
	}

	proof := make([]ProofNode, t.Depth)
	idx := index

	for level := 0; level < t.Depth; level++ {
		isRight := idx%2 == 1
		siblingIdx := idx + 1
		pathIndex := 0
		if isRight {
			siblingIdx = idx - 1
			pathIndex = 1
		}

		proof[level] = ProofNode{
			Element: t.Layers[level][siblingIdx],
			Index:   pathIndex,
		}

		idx = idx / 2
	}

	return proof
}

// VerifyProof verifies a Merkle proof for a given leaf hash.
func VerifyProof(leafHash, root fr.Element, proof []ProofNode) bool {
	current := leafHash

	for _, node := range proof {
		if node.Index == 1 {
			// Leaf is on right: hash(sibling, current)
			current = MiMCHash(&node.Element, &current)
		} else {
			// Leaf is on left: hash(current, sibling)
			current = MiMCHash(&current, &node.Element)
		}
	}

	return current.Equal(&root)
}

// ElementFromInt64 creates a field element from an int64.
func ElementFromInt64(v int64) fr.Element {
	var e fr.Element
	e.SetInt64(v)
	return e
}

// ElementFromBigInt creates a field element from a *big.Int.
func ElementFromBigInt(v *big.Int) fr.Element {
	var e fr.Element
	e.SetBigInt(v)
	return e
}

File: internal/static/public/merkle.js
// MiMC-BN254 Merkle tree — fixed-depth binary tree
// Matches gnark circuit Merkle proof verification exactly

import { mimcHash } from './mimc.js';

// Precompute zero hashes for each level of the tree.
// zeroHashes[0] = 0 (empty leaf)
// zeroHashes[i] = mimcHash(zeroHashes[i-1], zeroHashes[i-1])
function computeZeroHashes(depth) {
  const zeros = new Array(depth + 1);
  zeros[0] = 0n;
  for (let i = 1; i <= depth; i++) {
    zeros[i] = mimcHash(zeros[i - 1], zeros[i - 1]);
  }
  return zeros;
}

export class MerkleTree {
  constructor(root, leaves, depth, layers) {
    this.root = root;
    this.leaves = leaves;
    this.depth = depth;
    this.layers = layers;
  }

  // Build a fixed-depth Merkle tree from leaf values (BigInt field elements).
  // Leaves are hashed as mimcHash(key, value) before insertion.
  // If fewer leaves than 2^depth, remaining slots are padded with zero hashes.
  static fromEntries(entries, depth = 20) {
    const maxLeaves = 1 << depth;
    if (entries.length > maxLeaves) {
      throw new Error(`Too many entries: ${entries.length} > ${maxLeaves}`);
    }

    const zeroHashes = computeZeroHashes(depth);

    // Hash entries into leaves: mimcHash(key, value)
    const leafHashes = entries.map(([key, value]) => mimcHash(key, value));

    // Pad with zero leaf hashes
    while (leafHashes.length < maxLeaves) {
      leafHashes.push(zeroHashes[0]);
    }

    // Build tree bottom-up
    const layers = [leafHashes];
    let current = leafHashes;

    for (let level = 0; level < depth; level++) {
      const next = [];
      for (let i = 0; i < current.length; i += 2) {
        next.push(mimcHash(current[i], current[i + 1]));
      }
      layers.push(next);
      current = next;
    }

    return new MerkleTree(current[0], leafHashes, depth, layers);
  }

  // Build from pre-hashed leaves (BigInt values already hashed)
  static fromLeaves(leaves, depth = 20) {
    const maxLeaves = 1 << depth;
    if (leaves.length > maxLeaves) {
      throw new Error(`Too many leaves: ${leaves.length} > ${maxLeaves}`);
    }

    const zeroHashes = computeZeroHashes(depth);

    // Pad with zero hashes
    const padded = [...leaves];
    while (padded.length < maxLeaves) {
      padded.push(zeroHashes[0]);
    }

    const layers = [padded];
    let current = padded;

    for (let level = 0; level < depth; level++) {
      const next = [];
      for (let i = 0; i < current.length; i += 2) {
        next.push(mimcHash(current[i], current[i + 1]));
      }
      layers.push(next);
      current = next;
    }

    return new MerkleTree(current[0], padded, depth, layers);
  }

  // Get Merkle proof for leaf at index.
  // Returns { pathElements: BigInt[], pathIndices: number[] }
  // pathIndices[i] = 0 means leaf is on left, sibling on right
  // pathIndices[i] = 1 means leaf is on right, sibling on left
  // This matches the gnark circuit's Select logic exactly.
  getProof(index) {
    if (index < 0 || index >= this.leaves.length) {
      throw new Error(`Index out of range: ${index}`);
    }

    const pathElements = [];
    const pathIndices = [];
    let idx = index;

    for (let level = 0; level < this.depth; level++) {
      const isRight = idx % 2 === 1;
      const siblingIdx = isRight ? idx - 1 : idx + 1;

      pathElements.push(this.layers[level][siblingIdx]);
      pathIndices.push(isRight ? 1 : 0);

      idx = Math.floor(idx / 2);
    }

    return { pathElements, pathIndices };
  }

  // Update a leaf and recompute the tree
  updateLeaf(index, newLeafHash) {
    if (index < 0 || index >= this.leaves.length) {
      throw new Error(`Index out of range: ${index}`);
    }

    this.layers[0][index] = newLeafHash;
    let idx = index;

    for (let level = 0; level < this.depth; level++) {
      const pairIdx = idx % 2 === 0 ? idx + 1 : idx - 1;
      const parentIdx = Math.floor(idx / 2);
      const left = idx % 2 === 0 ? this.layers[level][idx] : this.layers[level][pairIdx];
      const right = idx % 2 === 0 ? this.layers[level][pairIdx] : this.layers[level][idx];
      this.layers[level + 1][parentIdx] = mimcHash(left, right);
      idx = parentIdx;
    }

    this.root = this.layers[this.depth][0];
    this.leaves = this.layers[0];
  }

  // Verify a Merkle proof
  static verifyProof(leafHash, root, pathElements, pathIndices) {
    let current = leafHash;
    for (let i = 0; i < pathElements.length; i++) {
      if (pathIndices[i] === 1) {
        // Leaf is on right: hash(sibling, current)
        current = mimcHash(pathElements[i], current);
      } else {
        // Leaf is on left: hash(current, sibling)
        current = mimcHash(current, pathElements[i]);
      }
    }
    return current === root;
  }
}
